service/history/timerQueueTaskExecutorBase.go

```diff
@@ -91,17 +95,28 @@ func (t *timerQueueTaskExecutorBase) executeDeleteHistoryEventTask(
 	defer func() { release(retError) }()
 
 	mutableState, err := loadMutableStateForTimerTask(weContext, task, t.metricsClient, t.logger)
-	if err != nil {
+	switch err.(type) {
+	case nil:
+		// continue the following step
+	case *serviceerror.NotFound:
+		return t.deleteHistoryBranch(task.BranchToken)
+	default:
 		return err
 	}
 
 	lastWriteVersion, err := mutableState.GetLastWriteVersion()
 	if err != nil {
 		return err
 	}
-	ok, err := verifyTaskVersion(t.shard, t.logger, namespace.ID(task.NamespaceID), lastWriteVersion, task.Version, task)
-	if err != nil || !ok {
-		return err
+	if ok := verifyTaskVersion(t.shard, t.logger, mutableState.GetNamespaceEntry(), lastWriteVersion, task.Version, task); !ok {
+		currentBranchToken, err := mutableState.GetCurrentBranchToken()
+		if err != nil {
+			return err
+		}
+		if !bytes.Equal(task.BranchToken, currentBranchToken) {
+			return t.deleteHistoryBranch(task.BranchToken)
+		}
+		return nil
```

Comment:
add comment for this case as well. this means version changed, but branch token keeps the same, we will do nothing here and wait for the new task (for new version) to delete history.